src/abstractdataframe/sort.jl

```diff
@@ -682,16 +701,117 @@ function Base.sort!(df::AbstractDataFrame, cols=All();
                     lt::Union{Function, AbstractVector{<:Function}}=isless,
                     by::Union{Function, AbstractVector{<:Function}}=identity,
                     rev::Union{Bool, AbstractVector{Bool}}=false,
-                    order::Union{Ordering, AbstractVector{<:Ordering}}=Forward)
+                    order::Union{Ordering, AbstractVector{<:Ordering}}=Forward,
+                    checkunique::Bool=false)
 
     # exclude AbstractVector as in that case cols can contain order(...) clauses
     if cols isa MultiColumnIndex && !(cols isa AbstractVector)
         cols = index(df)[cols]
     end
     ord = ordering(df, cols, lt, by, rev, order)
     _alg = Sort.defalg(df, ord; alg=alg, cols=cols)
+    checkunique && _perform_uniqueness_checks(df, cols, lt, by, order)
     return sort!(df, _alg, ord)
 end
 
-Base.sort!(df::AbstractDataFrame, a::Base.Sort.Algorithm, o::Base.Sort.Ordering) =
+function Base.sort!(df::AbstractDataFrame, a::Base.Sort.Algorithm,
+                    o::Base.Sort.Ordering, checkunique::Bool=false)
+    if checkunique
+        is_complex(o) &&
+            throw(ArgumentError("Orderings created with lt or by set to " *
+                                "something other than isless and identity " *
+                                "respectively, along checkunique=true are not supported."))
+        allunique(df, col_idxs) ||
+            throw(ArgumentError("Non-unique elements found. Multiple orders are valid."))
+    end
     permute!(df, _sortperm(df, a, o))
+end
+
+#= Functions to verify if an Ordering has user-defined by or lt functions
+The complexity checks exploit Base.Order's way of constructing Order objects.
+DirectOrdering is by definition an order that uses isless and identity as it's
+lt and by functions, so they are not complex. The By and Lt types have
+attributes storing the defined by and lt functions respectively, simple identity
+checks are enough.  ReverseOrderings wrap another type of ordering, so we
+perform the check on the wraped type.
+=#
+is_complex(o::Order.DirectOrdering) = false
+is_complex(o::By) = o.by !== identity
+is_complex(o::Lt) = o.lt !== isless
+is_complex(o::ReverseOrdering) = is_complex(o.fwd)
```

Comment:
I have re-checked `Base.Sort.Ordering` and its subtypes are:
```
julia> subtypes(Base.Sort.Ordering)
8-element Vector{Any}:
 Base.Order.By
 Base.Order.ForwardOrdering
 Base.Order.Lt
 Base.Order.Perm
 Base.Order.ReverseOrdering
 DataFrames.DFPerm
 DataStructures.FasterForward
 DataStructures.FasterReverse
```

I think that the one that needs to be added for sure is `DataFrames.DFPerm`.

These `Faster*` types probably are also easy to handle.

Reply:
Ok, `DFPerm` is easy enough. I'll have to take a look at `DataStructures`' code to handle their types

Reply:
I think `DataStructures.*` should be also easy 😄.
(and now you see the "rabbit hole" effect - this is, unfortunately, typical - a simple PR usually has many corner cases we need to think about that take a lot of work)

But I hope that at least this improves your knowledge of `sort` design in Julia in general 😄.

Reply:
You were right, `DataStructures` were the simplest

This has definitely improved my knowledge of `sort` and even DataFrames in general, It has been a very rewarding experience